### PR TITLE
[Core][Channel] Drop all channels at the end of a session

### LINKED ISSUE #104: Channels aren't finalized when a session ends
Once a channel is created, the channel will be kept inside a static map of `ChannelFactoryBase`. But channels inside this map are not finalized when a session ends, which leads to the following problem:
1. Channels created in the previous session will be passed to the next session if two or more sessions are run continuously, e.g. the binding.
2. Some resources will not be released. When @xwkuang5 is working on issue #49, he uses zmq sockets to pass buffers among workers. If push combine channels are not destroyed, these zmq sockets will not be closed. And `finalize_global()` will be stuck because zmq context keeps waiting for all zmq sockets to be closed.

### PR DESCRIPTION
**Pull request by zzxx-husky**
Fix #104.

Use session_local to finalize all the channels inside `finalize_global()`

### CODE REVIEW

**Comment by zzxx-husky** [comment]
Found a new problem:
Take `example/wc_mr.cpp` for example, the **source** of a push combine channel is an input format and the input format is a **local** variable in the `job()` function. When I try to destroy the push combine channel inside `finalize_local()` (`job()` function has finished),  troubles come because push combine channel tries to unregister the source.

File: core/channel/push_combined_channel.hpp [hunk 1]
```diff
@@ -45,7 +45,8 @@ class PushCombinedChannel : public Source2ObjListChannel<DstObjT> {
     ~PushCombinedChannel() override {
         ShuffleCombinerFactory::remove_shuffle_combiner(this->channel_id_);
 
-        this->src_ptr_->deregister_outchannel(this->channel_id_);
+        // TODO: what if src has already been released
+        // this->src_ptr_->deregister_outchannel(this->channel_id_);
```
Comment (zzxx-husky) [inline core/channel/push_combined_channel.hpp]: Don't know how to solve the problem at the moment, so I just turn this line into a comment.

File: core/channel/push_combined_channel.hpp [hunk 2]
```diff
@@ -45,7 +45,8 @@ class PushCombinedChannel : public Source2ObjListChannel<DstObjT> {
     ~PushCombinedChannel() override {
         ShuffleCombinerFactory::remove_shuffle_combiner(this->channel_id_);
 
-        this->src_ptr_->deregister_outchannel(this->channel_id_);
+        // TODO: what if src has already been released
```
Comment (kygx-legend) [inline core/channel/push_combined_channel.hpp]: `TODO(who)`

File: core/context.cpp [hunk 3]
```diff
@@ -22,13 +22,15 @@
...
 
+#include "base/session_local.hpp"
 #include "core/config.hpp"
 #include "core/coordinator.hpp"
 #include "core/hash_ring.hpp"
 #include "core/mailbox.hpp"
 #include "core/utils.hpp"
 #include "core/worker_info.hpp"
 
+
```
Comment (kygx-legend) [inline core/context.cpp]: Remove this new added line.

File: core/channel/push_combined_channel.hpp (hunk 2 above)
Comment (zzxx-husky) [inline core/channel/push_combined_channel.hpp]: I think I need a solution to update this PR. This line should not be commented. I just marked the problem here.

File: core/channel/channel_factory_base.cpp [hunk 4]
```diff
@@ -17,10 +17,14 @@
 #include <string>
 #include <unordered_map>
 
+#include "base/session_local.hpp"
+
 namespace husky {
 
 thread_local int ChannelFactoryBase::default_channel_id = 0;
 thread_local std::unordered_map<std::string, ChannelBase*> ChannelFactoryBase::channel_map;
```
Comment (ddmbr) [inline core/channel/channel_factory_base.cpp]: How about `std::unordered_map<std::string, std::unique_ptr<ChannelBase>>`? No more pain.

Comment (ddmbr) [inline core/channel/channel_factory_base.cpp]: @Yuzhen11

Comment (zzxx-husky) [inline core/channel/channel_factory_base.cpp]: Em ... still need someone to clear the map.

**Comment by kygx-legend** [comment]
LGTM

File: core/channel/channel_factory_base.cpp (hunk 4 above)
Comment (Yuzhen11) [inline core/channel/channel_factory_base.cpp]: Yes, I think we can use unique_ptr here.

Comment (kygx-legend) [inline core/channel/channel_factory_base.cpp]: @zzxx-husky, are you going to modify this?

**Comment by kygx-legend** [comment]
We may close this PR since it's included in another?

File: core/channel/channel_factory_base.cpp (hunk 4 above)
Comment (zzxx-husky) [inline core/channel/channel_factory_base.cpp]: @kygx-legend I just noticed this >_<. I think there will be lots of modifications if I change to `std::unique_ptr`? So maybe we fix this in a new PR?

Comment (kygx-legend) [inline core/channel/channel_factory_base.cpp]: Sure. Then close this PR now?

Comment (zzxx-husky) [inline core/channel/channel_factory_base.cpp]: Sure. OK to close this PR.

**Pull request closed without merging**